apigee/registry
# Implement Linting of Protos with API Linter (Plugin, Integration, Unit tests)

## Summary
In light of creating a method to specify style guides in the registry, it makes sense for us to have built-in support for API Linter. API-linter lints Protobuf specs according to the AIPs (https://aip.dev/), and was one of the main motivations for creating style guides in the registry in the first place.

This pull request implements the ability to specify style guides that contain rules from API linter, and have them be linted when computing conformance reports.

## Test
First, test using the description of this PR and ensure that the Spectral procedure still works correctly: https://github.com/apigee/registry/pull/304

Then,
- Have registry running on `localhost:8080`.
- Ensure that https://github.com/googleapis/googleapis is extracted to your Desktop.
- Ensure that the following style guide is located under `~/styleguide.yaml`:
```yaml
id: apilinterstyleguide
mime_types:
  - application/x.protobuf+zip
guidelines:
  - id: Naming
    display_name: Ensure Java multiple files is true
    rules:
      - id: TrademarkedNames
        description: This rule enforces trademarked names in public comments are not abbreviated and follow the trademark owner’s branding style, as mandated in AIP-192.
        linter: api-linter
        linter_rulename: core::0192::trademarked-names
        severity: 1
        doc_uri: https://linter.aip.dev/192/trademarked-names
    status: 2
linters:
  - name: api-linter
    uri: https://github.com/googleapis/api-linter
```

- Add the following two lines anywhere in the function with signature `func (task *computeConformanceTask) storeConformanceReport` in `conformance.go`:
```go
output, _ := json.Marshal(conformanceReport)
fmt.Println(string(output))
```

- Now, run the following series of bash commands:
```bash
export COMMIT=`(cd ~/Desktop/googleapis; git rev-parse HEAD)`

apg registry delete-project --name projects/protos

apg registry create-project --project_id protos --insecure --address localhost:8080

registry upload bulk protos --project-id protos ~/Desktop/googleapis --base-uri https://github.com/googleapis/googleapis/blob/$COMMIT

registry upload styleguide ~/styleguide-apilinter.yaml --project-id=protos

registry compute conformance projects/protos/locations/global/apis/google-storage/versions/v1/specs/protos.zip
```

The output on the command line should be the correct conformance report for the spec, which only reports the `core::0192::trademarked-names` which was specified in the style guide.
